File: leetcode/Python/non-negative-integers-without-consecutive-ones.py
class Solution:
    def findIntegers(self, n: int) -> int:
        sol, prev, dp = 0, 0, [1,2]
        for i in range(2,32):
            dp.append(dp[i-1]+dp[i-2])
        for i in range(30,-1,-1):
            if n&(1<<i) > 0:
                sol+=dp[i]
                if prev>0:
                    return sol
                prev=1
            else:
                prev=0
        return sol+1

File: leetcode/C++/non-negative-integers-without-consecutive-ones.cpp
class Solution {
public:
    int findIntegers(int n) {
        int sol=0, dp[32]={1,2}, prev=0;
        for(int i=2; i<32; ++i) dp[i]=dp[i-1]+dp[i-2];
        for(int i=30; i>=0; --i){
            if(n&(1<<i)){
                sol+=dp[i];
                if(prev) return sol;
                prev=1;
            }
            else prev=0;
        }
        return sol+1;
    }
};